lib/plugins/lindat_auth/__init__.py

```diff
@@ -1,163 +1,242 @@
                 return self.anonymous_user()
+            user_d = self._failover_auth.auth(self._db, username, password)
         else:
-            username = getenv('HTTP_EPPN') or getenv('HTTP_PERSISTENT_ID') or getenv('HTTP_MAIL')
-            if username is None or username == '':
-                return self.anonymous_user()
-            first_name = getenv('HTTP_GIVENNAME')
-            surname = getenv('HTTP_SN')
-            if not first_name and not surname:
-                full_name = getenv('HTTP_DISPLAYNAME') or getenv('HTTP_CN')
-                first_name, surname = self.parse_full_name(full_name)
-            cols = ('id', 'user', 'pass', 'firstName', 'surname')
-            cursor = self.db_conn.cursor()
-            cursor.execute("SELECT %s FROM user WHERE user = %%s" % ','.join(cols), (username, ))
-            row = cursor.fetchone()
-            if not row:
-                cursor.execute("INSERT INTO user (user, firstName, surname) VALUES (%s, %s, %s)",
-                               (username, first_name, surname))
-            else:
-                cursor.execute("UPDATE user SET firstName=%s, surname=%s WHERE user=%s",
-                               (first_name, surname, username))
-            cursor.execute("SELECT %s FROM user WHERE user = %%s" % ','.join(cols),
-                           (username, ))
-            row = cursor.fetchone()
-            row = dict(zip(cols, row)) if row else {}
-            cursor.close()
-            if 'id' in row:
-                return dict(id=row['id'],
-                            user=row['user'],
-                            fullname='%s %s' % (row['firstName'], row['surname']))
+            user_d = self._auth(plugin_api)
+
+        if user_d is not None:
+            user_id = int(user_d["id"])
+            return {
+                'id': user_id,
+                'user': user_d.get("username", "unknown"),
+                'fullname': user_d.get("fullname", "Mr. No Name")
+            }
+
         return self.anonymous_user()
 
+    def get_logout_url(self, return_url=None):
+        return self._logout_url
+
     def logout(self, session):
-        self.sessions.delete(session)
+        self._sessions.delete(session)
         session.clear()
 
     def permitted_corpora(self, user_id):
-        return self.corplist
+        # TODO(jm) based on user_id
+        return self._public_corplist
 
     def is_administrator(self, user_id):
-        """
-        Tests whether the current user's name belongs to the 'administrators' group
-        """
-        return user_id in self.admins
+        # TODO(jm)
+        return False
 
     def logout_hook(self, plugin_api):
         plugin_api.redirect('%sfirst_form' % (plugin_api.root_url,))
 
-    def get_restricted_corp_variant(self, corpus_name):
-        return corpus_name
+    def _new_user_id(self):
+        return self._db.incr(FederatedAuthWithFailover.RESERVED_USER)
```

Comment:
Is RESERVED_USER a db field to increment or a list of "protected user names"

Reply:
no it is a number  (in redis a string anyway)

Reply:
incr returns the value of the key which is a number - incr atomically increases it so it is thread safe